lib/commons/aria/roles.js

```diff
@@ -204,3 +202,248 @@ aria.implicitRole = function(node) {
 		ariaAttributes
 	).shift();
 };
+
+/**
+ * Get roles applied to a given node
+ * @method getRoleSegments
+ * @memberof axe.commons.aria
+ * @instance
+ * @param {HTMLElement} node HTMLElement
+ * @return {Array<String>} return an array of roles applied to the node, if no roles, return an empty array.
+ */
+aria.getRoleSegments = function(node) {
+	let roles = [];
+	if (node && node.hasAttribute('role')) {
+		roles = roles.concat(node.getAttribute('role').split(' '));
+	}
+	if (node && node.hasAttributeNS('http://www.idpf.org/2007/ops', 'type')) {
+		let rawEpubRoles = node.getAttributeNS(
+			'http://www.idpf.org/2007/ops',
+			'type'
+		);
+		let epubRoles = rawEpubRoles.split(' ');
+		roles = roles.concat(
+			epubRoles.map(role => {
+				return `doc-${role}`;
+			})
+		);
+	}
+	return roles;
+};
+
+/**
+ * Get roles applicable and or allowed for a given node/ HTMLElement
+ * @method getApplicableRolesForNode
+ * @instance
+ * @param {HTMLElement} node HTMLElement
+ * @return {Array<String>} return an array of roles applicable to the node, if no roles, return an empty array.
+ * @private
+ */
+function getApplicableRolesForNode(node) {
+	const tagName = node.tagName.toUpperCase();
+	const nodeAttrs = Array.from(node.attributes).map(a => a.name.toUpperCase());
+
+	const rolesTable = axe.commons.aria.lookupTable.role;
+	return Object.keys(axe.commons.aria.lookupTable.role).reduce((out, role) => {
+		const roleValue = rolesTable[role];
+		if (
+			roleValue.allowedElements &&
+			Array.isArray(roleValue.allowedElements) &&
+			roleValue.allowedElements.length
+		) {
+			// enumerate allowed elements and extract matching tagName
+			roleValue.allowedElements.forEach(el => {
+				// string
+				if (typeof el === 'string') {
+					if (el.toUpperCase() === tagName) {
+						out.push(role);
+					}
+					return;
+				}
+				// object - match tag in allowedElement
+				if (el.tagName.toUpperCase() === tagName) {
+					if (!el.attributes) {
+						out.push(role);
+						return;
+					}
+					el.attributes.forEach(attr => {
+						// check if attribute exsits
+						if (nodeAttrs.includes(attr.name.toUpperCase())) {
+							// check if value matches
+							if (attr.value === '') {
+								out.push(role);
+								return;
+							}
+							const nodeAttrValue = node.getAttribute(attr).toUpperCase();
+							if (attr.value.toUpperCase() === nodeAttrValue) {
+								out.push(role);
+								return;
+							}
+						}
+					});
+				}
+			});
+		}
+		return out;
+	}, []);
+}
+
+/**
+ * validate if a given role is an allowed ARIA role.
+ * @method validateAllowedAriaRole
+ * @param {String} tagName tag name of a node
+ * @param {String} role aria role to check
+ * @return {Boolean} retruns true/false
+ * @private
+ */
+function validateAllowedAriaRole(node, tagName, role) {
+	/**
+	 * Collection of functions, which evaluate if a given ROLE is allowed on a specified type of ELEMENT.
+	 * @type {Object}
+	 */
+	const elementCheck = {
```

Comment:
It's hard to tell what these things do. Some of them look like they could have been handled by the allowedElements property - can you add some tests that show how each of these is supposed to work? You may want to hoist this out of the function too.

Reply:
Not going to move this out of a function, as I do not see any re-use for this just yet. Have added tests.